### PR TITLE
istio-envoy-1.18.2: an Envoy build with Istio plugins

### PR DESCRIPTION
**Pull request by tcnghia**
Adding Istio's envoy build. It is based on upstream Envoy with additional policy and telemetry plugins.

#### For new package PRs only

- [ ] This PR is marked as fixing a pre-existing package request bug
  - [ ] Alternatively, the PR is marked as related to a pre-existing package request bug, such as a dependency
- [x] REQUIRED - The package is available under an OSI-approved or FSF-approved license
- [x] REQUIRED - The version of the package is still receiving security updates

### CODE REVIEW

File: istio-envoy.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      # We need to stick to gcc 12 for now, envoy doesn't build with gcc >= 13
+      - gcc-12-default
+
+pipeline:
+  - uses: git-checkout
+    with:
+      repository: https://github.com/istio/proxy
+      tag: ${{package.version}}
+      expected-commit: 3c27a1b0cf381ca854ccc3a2034e88c206928da2
+      destination: proxy
```
Comment (joshrwolf) [inline istio-envoy.yaml:45]:
```suggestion
```

nit: can we simplify this and just use the default `/home/build`?

File: istio-envoy.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  - runs: |
+      export JAVA_HOME=/usr/lib/jvm/java-11-openjdk
+      mkdir -p .cache/bazel/_bazel_root
+
+      cd proxy
+      echo "build --config=clang" >> user.bazelrc
+
+      # https://github.com/istio/proxy/commit/3c27a1b0cf381ca854ccc3a2034e88c206928da2 updated
+      # the Envoy commit without updating its checksum.
+      sed -i s/0a932fd090feed587d82b4b943a1032a5a424587f65455a37236cffefc8648ef/8795a73871ce8bd4138191f746d4f1f5d0935d5e65b2b6f1803f966f13065211/g WORKSPACE
```
Comment (joshrwolf) [inline istio-envoy.yaml:56]: will this change when the package version bumps?

**Comment by joshrwolf** [comment]
I don't know enough about enovy and it's plugin ecosystem, but is there any way to build/package these separately and load them at runtime?

**Comment by tcnghia** [comment]
> I don't know enough about enovy and it's plugin ecosystem, but is there any way to build/package these separately and load them at runtime?

Maybe there is, but we don't manage how to run this. The `proxy-agent` is the nanny for the proxy (which I will build separately) and it runs the proxy without much room for customization.

File: istio-envoy.yaml (hunk 1 above)
Comment (tcnghia) [reply istio-envoy.yaml]: Done!

I blame @rawlingsj because I copy-pasta-ed from the envoy.yaml :D

**Review by joshrwolf — DISMISSED** [review]

File: istio-envoy.yaml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      # When patching to 1.18.2, the PR
+      # https://github.com/istio/proxy/commit/3c27a1b0cf381ca854ccc3a2034e88c206928da2 updated
+      # the Envoy commit without updating its checksum.
+      sed -i s/0a932fd090feed587d82b4b943a1032a5a424587f65455a37236cffefc8648ef/8795a73871ce8bd4138191f746d4f1f5d0935d5e65b2b6f1803f966f13065211/g WORKSPACE
+
+      bazel build --verbose_failures -c opt envoy
+
+      # Istio proxy-agent expects Envoy here
+      # See https://github.com/istio/istio/blob/1.18.0/pkg/config/constants/constants.go#L59-L60
+      mkdir -p ${{targets.destdir}}/usr/local/bin/
```
[the remark this answers is not in the document]
Comment (tcnghia) [reply istio-envoy.yaml:61]: Thanks for the pointer! Done

**Review by rawlingsj — APPROVED** [review]

**Pull request merged**